core/store/src/trie/mod.rs

```diff
@@ -1711,17 +1712,51 @@ impl Trie {
         &self,
         boundary_account: &AccountId,
         retain_mode: RetainMode,
-        opts: AccessOptions,
-    ) -> Result<StateRoot, StorageError> {
-        let mut trie_update = TrieStorageUpdate::new(&self);
-        let root_node = self.move_node_to_mutable(&mut trie_update, &self.root, opts)?;
-        trie_update.retain_split_shard(boundary_account, retain_mode, opts);
-        #[cfg(test)]
-        {
-            self.memory_usage_verify(&trie_update, GenericNodeOrIndex::Updated(root_node.0));
+    ) -> Result<TrieChanges, StorageError> {
+        if self.memtries.is_some() {
+            // Get child trie_changes for all child memtries
+            let children_memtrie_changes = self
+                .children_memtries
+                .iter()
+                .map(|(shard_uid, memtrie)| {
+                    let inner_guard = memtrie.read().unwrap();
+                    let mut trie_update =
+                        inner_guard.update(self.root, TrackingMode::None).unwrap();
+                    trie_update.retain_split_shard(
+                        boundary_account,
+                        retain_mode,
+                        AccessOptions::DEFAULT,
+                    );
+                    (*shard_uid, trie_update.to_memtrie_changes_only())
+                })
+                .collect::<HashMap<_, _>>();
+
+            // Get trie_update for memtrie
+            let guard = self.memtries.as_ref().unwrap().read().unwrap();
+            let mut recorder =
+                self.recorder.as_ref().map(|recorder| recorder.write().expect("no poison"));
+            let tracking_mode = match &mut recorder {
+                Some(recorder) => TrackingMode::RefcountsAndAccesses(recorder.deref_mut()),
+                None => TrackingMode::Refcounts,
+            };
+            let mut trie_update = guard.update(self.root, tracking_mode)?;
+            trie_update.retain_split_shard(boundary_account, retain_mode, AccessOptions::DEFAULT);
+
+            let mut trie_changes = trie_update.to_trie_changes();
+            trie_changes.children_memtrie_changes = children_memtrie_changes;
+            Ok(trie_changes)
+        } else {
```

Comment:
Why is there a difference for memtrie and no memtrie handling? Shouldn't we also do some recording?

Reply:
Right, so here, we are using the same function for two use cases, 1) generation and 2) validation. 

The generation code path calls this function in resharding manager where this PR is making the bulk of the changes.

The validation code path is in `chunk_validation.rs` which calls this function from partial trie/state witness proof which use this else path (which is the original code).